infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LocationTrackService.kt

```diff
@@ -448,41 +459,48 @@ class LocationTrackService(
     ): List<LocationTrackDuplicate> {
-        val originalAndAlignment = getWithAlignmentOrThrow(publicationState, id)
-        val duplicates = dao.fetchDuplicateVersions(id, publicationState).map(dao::fetch)
-
-        val duplicateMValues = duplicates.map { duplicate ->
-            duplicate.alignmentVersion?.let { alignmentVersion ->
-                alignmentDao.fetch(alignmentVersion).firstSegmentStart?.let(
-                    originalAndAlignment.second::getClosestPointM
-                )?.first
-            }
-        }
-        return duplicates
-            .mapIndexed { index, d -> index to d }
-            .sortedWith { a, b -> compareValues(duplicateMValues[a.first], duplicateMValues[b.first]) }
-            .map { (_, track) ->
-                LocationTrackDuplicate(track.id as IntId, track.trackNumberId, track.name, track.externalId)
+        val switchJoints = collectSwitchJoints(track, alignment)
+
+        val markedDuplicateVersions = dao.fetchDuplicateVersions(track.id as IntId, publicationState)
+        val tracksLinkedThroughSwitch = switchDao.findLocationTracksLinkedToSwitches(publicationState, track.switchIds)
+            .map(LayoutSwitchDao.LocationTrackIdentifiers::rowVersion)
+
+        return (markedDuplicateVersions + tracksLinkedThroughSwitch)
+            .asSequence()
+            .distinct()
+            .map(::getWithAlignmentInternal)
+            .filter { (duplicateTrack, _) -> duplicateTrack.id != track.id }
+            .flatMap { (duplicateTrack, duplicateAlignment) ->
+                getLocationTrackDuplicatesByJoint(track.id, switchJoints, duplicateTrack, duplicateAlignment)
             }
+            .sortedWith(compareBy({ it.first }, { it.second.name }))
+            .map { (_, duplicate) -> duplicate }
+            .toList()
     }
 
     private fun getDuplicateOf(
         locationTrack: LocationTrack,
         publicationState: PublicationState,
-    ) = locationTrack.duplicateOf?.let { duplicateId ->
-        get(publicationState, duplicateId)?.let { dup ->
-            LocationTrackDuplicate(
-                duplicateId, dup.trackNumberId, dup.name, dup.externalId
-            )
+    ): LocationTrackDuplicate? = locationTrack.duplicateOf?.let { duplicateOfId ->
+        getWithAlignment(publicationState, duplicateOfId)?.let { (duplicateOfTrack, duplicateOfAlignment) ->
+            val alignment = alignmentDao.fetch(locationTrack.getAlignmentVersionOrThrow())
+            // Note arguments: here the input "locationTrack" is the duplicate of the other ("main") one
+            getLocationTrackDuplicatesByJoint(
+                mainTrackId = duplicateOfId,
+                mainTrackJoints = collectSwitchJoints(duplicateOfTrack, duplicateOfAlignment),
+                duplicateTrack = locationTrack,
+                duplicateAlignment = alignment,
+            ).first().second // There has to at least one found, since we know the duplicateOf is set
```

Comment:
Onkohan validi tilanne että jostain syystä löytyisikin yli yksi? 🤔 Lähinnä mietin pitäisikö tässä käyttää `single()`:ä `first()`:in sijasta